[maintenance] Update some pointers to smart pointers

lrusak opened the pull request:
This PR continues my quest to remove `SAFE_DELETE` and `system.h`. I could just replace the `SAFE_DELETE` macro but I think updating the code is better.

All the changes are quite trivial and there isn't anything I'm really concerned about but a second set of eyes would be good.

Most changes just involve using a `unique_ptr`

lrusak committed:
CPlayerCoreConfig: remove use of SAFE_DELETE

lrusak committed:
CVideoPlayerAudio: make m_pAudioCodec a unique_ptr

lrusak committed:
CVideoPlayerSubtitle: remove unused member m_pSubtitleStream

lrusak committed:
CVideoPlayerSubtitle: make m_pSubtitleFileParser a unique_ptr

lrusak committed:
CVideoPlayerSubtitle: make m_pOverlayCodec a unique_ptr

lrusak committed:
CDVDVideoCodecAndroidMediaCodec: make m_bitstream a unique_ptr

lrusak committed:
CFile: make m_pFile a unique_ptr

lrusak committed:
CFile: make m_pBuffer a unique_ptr

lrusak committed:
CFile: make m_bitStreamStats a unique_ptr

lrusak committed:
CFileStream: make m_file a unique_ptr

lrusak committed:
CVideoPlayer: make m_pDemuxer a unique_ptr

lrusak committed:
CVideoPlayer: make m_pCCDemuxer a unique_ptr

garbear commented on xbmc/cores/VideoPlayer/VideoPlayerAudio.cpp:
```diff
@@ -53,7 +53,7 @@ CVideoPlayerAudio::CVideoPlayerAudio(CDVDClock* pClock, CDVDMessageQueue& parent
 {
   m_pClock = pClock;
-  m_pAudioCodec = NULL;
+  m_pAudioCodec = nullptr;
```
No need to call `std::unique_ptr::operator=` here

lrusak replied:
Thanks, updated

garbear commented on xbmc/cores/playercorefactory/PlayerCoreConfig.h:
```diff
@@ -104,16 +99,19 @@ class CPlayerCoreConfig
       return nullptr;
 
+    if (!pPlayer)
+      return nullptr;
```
If you want to go further, `CPlayerCoreConfig::CreatePlayer()` should return a unique_ptr

lrusak replied:
I've actually done this already and will prepare a PR soon 😄

garbear commented on xbmc/filesystem/File.cpp:
```diff
@@ -291,15 +285,15 @@ bool CFile::Open(const CURL& file, const unsigned int flags)
       }
     }
-    m_pFile = CFileFactory::CreateLoader(url);
+    m_pFile.reset(CFileFactory::CreateLoader(url));
```
It might be more clear to user operator= instead of reset, e.g.
```c++
m_pFile = std::move(CFileFactory::CreateLoader(url));
```
(not sure if the std::move is required but I think it is)

lrusak replied:
I was actually unaware of this, thanks for pointing that out!

arnova replied:
Yes, std::move is clearer here. Same for "m_pDemuxer.reset(CDVDFactoryDemuxer::CreateDemuxer(m_pInputStream));" above.

lrusak replied:
it's not possible to use the `operator=()` in both instances as `.reset()` initialises the `unique_ptr` (that's why I was unaware, it's not possible :upside_down_face:).

```
/home/lukas/Documents/git/xbmc/xbmc/filesystem/File.cpp:1205:53: error: no match for ‘operator=’ (operand types are ‘std::unique_ptr<XFILE::IFile>’ and ‘std::remove_reference<XFILE::IFile*>::type’ {aka ‘XFILE::IFile*’})
 1205 |   m_file = std::move(CFileFactory::CreateLoader(url));
      |                                                     ^
```

Both `CFileFactory::CreateLoader` and `CDVDFactoryDemuxer::CreateDemuxer` return raw pointers which are used to initialise the unique_ptr with `.reset()`

If they returned a `unique_ptr` this would be possible but I think that's outside the scope of this PR and can be included in a future PR.

garbear replied:
ok got it. future PR sounds good if you get to it

garbear commented on xbmc/filesystem/File.h:
```diff
@@ -103,11 +104,11 @@ class CFile
   static int DetermineChunkSize(const int srcChunkSize, const int reqChunkSize);
 
-  BitstreamStats* GetBitstreamStats() { return m_bitStreamStats; }
+  BitstreamStats* GetBitstreamStats() { return m_bitStreamStats.get(); }
```
I'm thinking that calling `.get()` conflicts with the idea of a wrapped pointer. We can probably return a `std::unique_ptr&`, and let the caller make the decision to unwrap only if needed.

Going further, would could return a `const std::unique_ptr&` and label this function const to improve const-correctness. This might be a new commit or PR if propagating const-correctness requires changes in calling code.

lrusak replied:
I considered going do the path of returning a `unique_ptr` but thought it would be too much bloat for this PR. I'll look into it again and see how much work it may be. If it's trivial I'll include the changes in this PR otherwise I can make a follow up PR.

garbear replied:
Returning a `unique_ptr&` should be trivial - find all compiler errors an add a `.get()`. Adding const-correctness is more work, and do whatever is easiest for you (this PR or a new one) - I can review either way

lrusak replied:
updated, it was trivial

garbear commented:
Done with first round of review, left some comments.

lrusak committed:
CFile: use default constructor

lrusak commented:
I've updated this to address some of the comments.

I've also added a commit to default initialise `CFile` I had to leave this in the cpp file as we do some forward declarations in the header and didn't want to move headers around.
